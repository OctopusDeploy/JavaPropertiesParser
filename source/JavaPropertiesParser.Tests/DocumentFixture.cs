using System;
using System.Collections.Generic;
using Assent;
using JavaPropertiesParser.Expressions;
using JavaPropertiesParser.Tests.TestUtils;
using JavaPropertiesParser.Utils;
using NUnit.Framework;

namespace JavaPropertiesParser.Tests
{
    public class DocumentFixture
    {
        [Test]
        public void CanReplaceValuesWhilePreservingFormatting()
        {
            var keyValues = new Dictionary<string, string>
            {
                { "KEY1", "The new value for key\n1" },
                { "key\\4", "The new value for key 4" },
                { "key5", "Key 5's\n\n\nnewvalue" },
                { "notarealkey", "Should not be found" },
                { "", "New value for the empty key" },
                { "key15", "" }
            };

            ITopLevelExpression Mutator(ITopLevelExpression expr)
            {
                switch (expr)
                {
                    case null:
                        throw new ArgumentNullException(nameof(expr));

                    case KeyValuePairExpression pair:
                        var key = pair.Key?.Text?.LogicalValue ?? "";
                        if (keyValues.TryGetValue(key, out var value))
                        {
                            return new KeyValuePairExpression(
                                pair.Key,
                                pair.Separator,
                                new ValueExpression(
                                    new StringValue(
                                        value,
                                        Encode.Value(value)
                                    )
                                )
                            );
                        }
                        else
                        {
                            return expr;
                        }

                    default:
                        return expr;
                }
            }
            
            var input = ResourceUtils.ReadEmbeddedResource("multiple-pairs-and-comments.properties");
            var parsed = Parser.Parse(input);
            var updated = parsed.Mutate(Mutator);

            var actual = updated.ToString();
            this.Assent(actual, TestEnvironment.AssentConfiguration);
        }
    }
}