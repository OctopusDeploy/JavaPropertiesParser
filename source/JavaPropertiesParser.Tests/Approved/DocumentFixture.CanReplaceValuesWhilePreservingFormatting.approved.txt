# Test file
key1=value1
KEY1=The new value for key\n1
key\ 2:value2
key\3= value3
key\\4 :The new value for key 4
key\
5		: Key 5's\n\n\nnewvalue
key6
key7:
key8:\
value8

! Another comment
key\r\n\t\\\:9:value 9
key10\
:value\
10
:New value for the empty key
k\u00c8y12:valu\u00c8
key13==value13
key14 value14
key15	
key16:value\r\n\\16
key17:\
               \
  \
         value17
